### PR TITLE
Make username and password in SMTP driver optional

### PR DESCRIPTION
**Pull request by sfc-gh-pkommini**
Make username and password optional

Tests:

Tested HTTP Driver:
image

Tested SMTP Driver:
image

### CODE REVIEW

File: lambda_src/drivers/process_smtp.py [hunk 1]
```diff
@@ -1,16 +1,72 @@
...
 import smtplib
 import ssl
 from email.mime.multipart import MIMEMultipart
 from email.mime.text import MIMEText
+from typing import Dict, Tuple, Any
 
 from ..vault import decrypt_if_encrypted
 
 
+def parse_smtp_creds(auth_dict: Dict) -> Tuple[Any, Any, Any]:
```
Comment (sfc-gh-afedorov) [inline lambda_src/drivers/process_smtp.py:11]: could we be a little more prescriptive and also make sure we can't possibly confuse some other drivers' credentials to be re-used here? maybe e.g. add a `"driver": "smtp"` key or just force `smtp_host` to make sure another driver wouldn't re-use these credentials?

sooner or later we'll need to document this and having a single option would make things simpler there, as well. people shouldn't be guessing here, and docs should give them one right answer.

that said, do appreciate the cool next/map combo, of course :)

**Review by sfc-gh-afedorov — APPROVED** [review]
lgtm, minor feedback re: fewer options being better

File: lambda_src/drivers/process_https.py [hunk 2]
```diff
@@ -59,6 +58,7 @@ def process_row(
     req_headers.setdefault('User-Agent', 'GEFF 1.0')
     req_headers.setdefault('Accept-Encoding', 'gzip')
 
+    # We look for an auth header and if found we parse it from it's encoded format
```
[the remark this answers is not in the document]
Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_https.py:61]: ACK.

File: lambda_src/drivers/process_smtp.py [hunk 3]
```diff
@@ -27,6 +37,15 @@ def process_row(
     password = decrypt_if_encrypted(password)
     sender_email = sender_email or user
 
+    if auth:
+        auth_json = decrypt_if_encrypted(auth)
+        auth_dict = json.loads(auth_json)
+        user, password, host = parse_smtp_creds(auth_dict)
```
[the remark this answers is not in the document]
Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_smtp.py:43]: There is nothing to match the host with. As there is no host passed to the driver from the external function, so the threat model here is a bit different in that if the `auth` dict doesn't have an `SMTP_HOST` we reject the request.

File: lambda_src/drivers/process_smtp.py [hunk 4]
```diff
@@ -27,6 +37,15 @@ def process_row(
...
+        if not host:
+            raise ValueError(
+                'Credentials not pinned to an SMTP host! This can increase the risk of credential compromise.'
```
[the remark this answers is not in the document]
Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_smtp.py:46]: ACK.

File: lambda_src/drivers/process_https.py [hunk 5]
```diff
@@ -59,22 +58,29 @@ def process_row(
...
             loads(auth)
             if auth.startswith('{')
             else parse_header_dict(auth)
             if auth
             else {}
         )
+        auth_host = req_auth.get('host')
+
+        if not auth_host:
+            raise ValueError('\'auth\' is missing the \'SMTP_HOST\' key.')
```
Comment (sfc-gh-afedorov) [inline lambda_src/drivers/process_https.py:76]: - use `"` for strings that are shown to users
- the key is `'host'` here

File: lambda_src/drivers/process_https.py [hunk 6]
```diff
@@ -59,22 +58,29 @@ def process_row(
...
 
-        if 'host' in req_auth and not req_host:
-            req_host = req_auth['host']
+        # We reject the request if the 'auth' is not pinned to a host or doesn't match the pinned host
+        if auth_host != req_host:
```
Comment (sfc-gh-afedorov) [inline lambda_src/drivers/process_https.py:79]: keep `auth_host is Null` in this check?

Comment (sfc-gh-afedorov) [reply lambda_src/drivers/process_https.py:79]: that's the opposite of what we want -- if the auth host is missing, we want to send an un-authenticated https request

Comment (sfc-gh-afedorov) [reply lambda_src/drivers/process_https.py:79]: in general, can we keep refactoring of https in a separate PR? this one is supposed to be a change to the SMTP handler

File: lambda_src/drivers/process_smtp.py [hunk 7]
```diff
@@ -75,13 +94,17 @@ def process_row(
 
     try:
         result = smtpserver.sendmail(sender_email, recipients, message.as_string())
+    except ValueError as e:
```
[the remark this answers is not in the document]
Comment (sfc-gh-afedorov) [reply lambda_src/drivers/process_smtp.py:104]: could we put just that part into a try/except, then, and have a separate try for sendmail?

File: lambda_src/drivers/process_smtp.py [hunk 8]
```diff
@@ -27,6 +39,20 @@ def process_row(
     password = decrypt_if_encrypted(password)
     sender_email = sender_email or user
 
+    if auth:
+        auth_json = decrypt_if_encrypted(auth)
+        assert auth is not None
```
Comment (sfc-gh-afedorov) [inline lambda_src/drivers/process_smtp.py:44]: when would this be the case? iirc decrypt raises an exception if it cannot decrypt, and returns `None` only if `auth` is `None`

File: lambda_src/drivers/process_https.py (hunk 6 above)
Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_https.py:79]: ACK. I was making the pinning uniform for both. I thought this PR was related to our pinning discussion hence the changes related to that.

File: lambda_src/drivers/process_smtp.py (hunk 8 above)
Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_smtp.py:44]: I tried it make it uniform with the HTTPS driver but yeah will remove the assertion in both if that's the case.

File: lambda_src/drivers/process_https.py (hunk 6 above)
Comment (sfc-gh-afedorov) [reply lambda_src/drivers/process_https.py]: thanks, yes -- we want user / password to be optional in SMTP. we want to retain the optional auth in https

Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_https.py]: It's done. Let me know if the changes look okay. Opened a new PR for the HTTP changes.

Comment (sfc-gh-pkommini) [reply lambda_src/drivers/process_https.py]: Tested with secrets manager. 
image

**Pull request merged**